File: lib/core/services/department_head_guard.dart
// lib/core/services/department_head_guard.dart
//
// One department head per department, client side.
//
// The guarantee is the trigger `department_table_one_head` from migration
// 20240130000016. This file exists so the screens can say so in a sentence --
// before an account is created, before an OTP is emailed, and while the
// dropdown is still open -- rather than surfacing a rejected write.
//
// Three copies of the role rule exist and must agree:
//   * public.is_department_head_role(bigint)  -- migration 20240130000016
//   * isDepartmentHeadRole below
//   * isDepartmentHeadRole -- supabase/functions/_shared/department_head_guard.ts
// The database one is authoritative; it backs the trigger.

import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

/// Raw role strings that land on the department-head dashboard.
///
/// DEAN is in here on purpose. [roleFromString] in role_nav_config routes DEAN
/// and DEPARTMENT_HEAD to the same [UserRole.deptHead], so two of them over one
/// department is exactly the collision this rule prevents -- the app could not
/// tell which of the two owns the faculty roster.
const Set<String> kDepartmentHeadRoles = <String>{
  'DEPARTMENT_HEAD',
  'DEPARTMENT-HEAD',
  'DEPARTMENT HEAD',
  'DEAN',
};

/// Is this raw role string a department-head role?
bool isDepartmentHeadRole(String? raw) {
  if (raw == null) return false;
  final normalised = raw.trim().toUpperCase().replaceAll(RegExp(r'[\s_-]+'), '-');
  return normalised == 'DEPARTMENT-HEAD' || normalised == 'DEAN';
}

/// Pre-flight checks and wording for the one-head-per-department rule.
class DepartmentHeadGuard {
  const DepartmentHeadGuard._();

  /// The sentence every screen shows when the chair is taken.
  ///
  /// [headName] is omitted where the caller has no business naming the sitting
  /// head -- the registration screen, where the applicant is not staff yet.
  static String conflictMessage({String? headName, String? departmentName}) {
    final where = (departmentName != null && departmentName.trim().isNotEmpty)
        ? departmentName.trim()
        : 'that department';
    if (headName != null && headName.trim().isNotEmpty) {
      return '$headName is already the department head of $where. '
          'A department can only have one head — change that person\'s role '
          'first, or choose a different department.';
    }
    return 'A department head is already assigned to $where. '
        'A department can only have one head, so this role is not available '
        'for it. Pick another department, or contact the SAO office.';
  }

  /// Does [departmentId] already have a head?
  ///
  /// Calls the `department_has_head` RPC rather than reading department_table
  /// directly: the registration screen asks this while nobody is signed in,
  /// and the RPC is SECURITY DEFINER and hands back a boolean instead of staff
  /// rows.
  ///
  /// [excludeUserId] is the account being edited, so the sitting head does not
  /// block their own save.
  ///
  /// Returns false if the check itself cannot be reached. It is a pre-flight
  /// courtesy, not the guarantee -- the trigger still rejects the write and
  /// [describeConflictError] turns that into the same message. Blocking every
  /// role change because one RPC timed out would be the worse failure.
  static Future<bool> departmentHasHead({
    required SupabaseClient client,
    required Object departmentId,
    String? excludeUserId,
  }) async {
    try {
      final result = await client.rpc(
        'department_has_head',
        params: {
          'p_department_id': departmentId is int
              ? departmentId
              : int.tryParse(departmentId.toString()),
          'p_exclude_user_id': excludeUserId,
        },
      );
      return result == true;
    } on SocketException {
      debugPrint('[DEPT_HEAD] Vacancy check skipped: no connection.');
      return false;
    } catch (e) {
      // Most likely migration 20240130000016 has not been applied yet, so the
      // function does not exist. Loud in debug, silent for the user.
      debugPrint('[DEPT_HEAD] Vacancy check unavailable: $e');
      return false;
    }
  }

  /// Recognises the head conflict in an error raised by the database or
  /// returned by an edge function, and hands back the message to show.
  ///
  /// Both sources start with the same words: the trigger raises
  /// "Department head already assigned: ..." and the edge functions list that
  /// prefix among the messages they pass through untouched.
  static String? describeConflictError(Object error) {
    final message = error is PostgrestException
        ? error.message
        : error.toString();
    final marker = message.indexOf('Department head already assigned');
    if (marker == -1) return null;
    // Trim the leading marker so the sentence reads as prose in a dialog.
    return message.substring(marker).replaceFirst(
          RegExp(r'^Department head already assigned:\s*'),
          '',
        );
  }
}

File: supabase/functions/_shared/department_head_guard.ts
// supabase/functions/_shared/department_head_guard.ts
//
// One department head per department, server side.
//
// The guarantee is the trigger `department_table_one_head` from migration
// 20240130000016 -- it holds for every write, including a REST call made with
// the publishable key. This module exists so the two admin functions can say
// WHO is already sitting in the chair before they create an account, send an
// OTP or fire a notification email, instead of surfacing a failed insert.
//
// Four names count as a head, matching role_nav_config.roleFromString: all of
// them route to the same dept-head dashboard, so two of them over one
// department is exactly the collision being prevented.
//
// Three copies of this rule exist and must agree:
//   * public.is_department_head_role(bigint)  -- migration 20240130000016
//   * isDepartmentHeadRole below
//   * kDepartmentHeadRoles -- lib/core/services/department_head_guard.dart
// The database one is authoritative; it backs the trigger.

// deno-lint-ignore-file no-explicit-any

/**
 * Prefix every head-conflict message starts with.
 *
 * Both functions list it in their `safePrefixes`, so the sentence reaches the
 * admin verbatim instead of being flattened to "Operation failed". The
 * trigger's own exception message starts with the same words, which is what
 * lets a message raised by the database pass the same filter.
 */
export const HEAD_CONFLICT_PREFIX = 'Department head already assigned'

/** Is this role name one that lands on the department-head dashboard? */
export function isDepartmentHeadRole(roleName: unknown): boolean {
  const normalised = String(roleName ?? '')
    .trim()
    .toUpperCase()
    .replace(/[\s_-]+/g, '-')
  return normalised === 'DEPARTMENT-HEAD' || normalised === 'DEAN'
}

/**
 * The head currently sitting for [deptId], or null when the chair is free.
 *
 * [excludeUserId] is the account being edited, so re-saving the sitting head
 * does not report them as blocking themselves.
 */
export async function findDepartmentHead(
  client: any,
  deptId: number | string,
  excludeUserId?: string | null,
): Promise<{ userId: string | null; name: string } | null> {
  const { data, error } = await client
    .from('department_table')
    .select('user_id, roles:roles(id, Roles), user_info!user_id(first_name, last_name, account_status)')
    .eq('Department_name_ID', deptId)
  if (error) throw error

  for (const row of (data ?? [])) {
    if (excludeUserId && row.user_id === excludeUserId) continue

    // The embed comes back as an object for a to-one relationship and as an
    // array when PostgREST cannot prove it is to-one. Both shapes appear in
    // this project already, so both are handled.
    const roleRaw = Array.isArray(row.roles) ? row.roles[0] : row.roles
    if (!isDepartmentHeadRole(roleRaw?.Roles)) continue

    const info = Array.isArray(row.user_info) ? row.user_info[0] : row.user_info
    // A deleted account is not sitting in the chair.
    if (String(info?.account_status ?? '').toLowerCase() === 'deleted') continue

    const name = [info?.first_name, info?.last_name]
      .filter((part: unknown) => typeof part === 'string' && part.trim() !== '')
      .join(' ')
      .trim()

    return { userId: row.user_id ?? null, name: name || 'another account' }
  }

  return null
}

/** The sentence shown to the admin. Starts with [HEAD_CONFLICT_PREFIX]. */
export function describeHeadConflict(headName: string, deptName?: string | null): string {
  const where = deptName && deptName.trim() !== '' ? deptName : 'that department'
  return `${HEAD_CONFLICT_PREFIX}: ${headName} already holds the department head ` +
    `role for ${where}. A department can only have one head -- change that ` +
    `person's role first, or choose a different department.`
}

/**
 * Throws the readable conflict when [deptId] already has a head and
 * [roleName] would add another. No-op for every other role.
 */
export async function assertDepartmentHeadVacant(
  client: any,
  opts: {
    roleName: unknown
    deptId: number | string | null | undefined
    excludeUserId?: string | null
  },
): Promise<void> {
  if (!isDepartmentHeadRole(opts.roleName)) return
  if (opts.deptId === null || opts.deptId === undefined) return

  const head = await findDepartmentHead(client, opts.deptId, opts.excludeUserId)
  if (!head) return

  const { data: deptRow } = await client
    .from('department_name').select('d_name').eq('id', opts.deptId).limit(1)
  const deptName = deptRow && deptRow.length > 0 ? deptRow[0].d_name : null

  throw new Error(describeHeadConflict(head.name, deptName))
}
